include/net/tcp.h

```diff
@@ -1490,8 +1490,11 @@ static inline void tcp_slow_start_after_idle_check(struct sock *sk)
 	    ca_ops->cong_control)
 		return;
 	delta = tcp_jiffies32 - tp->lsndtime;
-	if (delta > inet_csk(sk)->icsk_rto)
+	if (delta > inet_csk(sk)->icsk_rto) {
+		tp->snd_cwnd_before_idle_restart =
```

Comment:
We should try to find a cleaner way to update this variable without having to introduce changes to the regular TCP stack.

Reply:
Looking through the code, ```snd_cwnd_before_idle_restart``` is only used when ```ecf_optim``` is set. That latter one defaults to 0 however. Without having read the paper yet, I'm not sure if we need it.

Reply:
> Looking through the code, `snd_cwnd_before_idle_restart` is only used when `ecf_optim` is set. That latter one defaults to 0 however. Without having read the paper yet, I'm not sure if we need it.

I have read the paper again and `ecf_optim` is not mentioned as part of the scheduling algorithm. I conclude that this could only be a remainder of the author's evaluation to prove the effect of resetting the CWND by simply turning off their proposed behavior. Thus, we can remove this module parameter completely.

Reply:
> We should try to find a cleaner way to update this variable without having to introduce changes to the regular TCP stack.

@cpaasch I totally agree with you. Do you have any suggestions for properly identifying this event from a scheduling perspective?

Reply:
From what I understand, if you remove the ```ecf_optim```, there isn't a need anymore for this code here in ```include/net/tcp.h```.

Reply:
No, we still need to save the cwnd value. I have now moved it to ecfsched_priv. This is actually the main idea of the proposed algorithm to avoid slow-start (in the estimation) again.

Meanwhile I did some major cleanup of the oringinal code to make it more readable. The only thing i still did not solve is how to detect when the subflows cwnd was reset due to a idle-phase.

Reply:
Why do we need to save it? AFAICS, it is only read from when ```ecf_optim``` is set, no?

Reply:
I sugguest to completly remove the ecf_optim module paramater and by default always use max(besttp->snd_cwnd, ecf_p->snd_cwnd_before_idle_restart) as it is proposed in the paper.

We need to save the cwnd value because TCP resets it when the subflow is idling for a RTO. In that case, we simple estimate the completion time with the previous ly recorded value in order not to underestimate the subflows performance.

Reply:
Ok, I got confused because ```ecf_optim``` was initialized to 0. So, the intention is to default-enable ```ecf_optim``` ?

If a subflow is idle such that the snd_cwnd gets reset, we don't know anymore what the capacity of that path is.
In that case, can we really assume that it is going to be the same capacity as before the idle-period? If the assumption in the network is that the capacity remains the same over idle periods, then one could simply disable restart-after-idle globally, no?
Further, during these idle-periods the congestion-window is only decreasing gradually by a factor of 2 per RTO.

Looking at the paper [1], I don't see a reference to the restart-after-idle part. Maybe you can start an e-mail thread with the authors (CC me and Matt) to clarify this part ?

I'm hesitant to introduce this kind of change to the TCP-stack without being convinced that it is really necessary.


[1] https://www.repository.cam.ac.uk/bitstream/handle/1810/279113/ecf_extended_abstract_rev1.pdf?sequence=1

Reply:
> Looking at the paper [1], I don't see a reference to the restart-after-idle part.

@cpaasch I just noticed that we are looking at two different versions of the same paper. Your are looking at the one that was published at SIGMETRICS’17 [1] and I am looking at the longer one published at CoNEXT ’17 [2]. The later describes the effect of idle subflows in detail in section 3.2

> Ok, I got confused because ecf_optim was initialized to 0. So, the intention is to default-enable ecf_optim ?

Yes. At least this is the way I understand the conclusion given in [2] and therefore would suggest to remove it completely. Because when you choose to use ECF instead of BLEST or the default scheduler, this is the feature you actually expect to use, or don't you?

> If a subflow is idle such that the snd_cwnd gets reset, we don't know anymore what the capacity of that path is.
In that case, can we really assume that it is going to be the same capacity as before the idle-period? If the assumption in the network is that the capacity remains the same over idle periods, then one could simply disable restart-after-idle globally, no?

Good point. If we disable restart-after-idle globally (in the case of almost constants capacities) we may prevent the performance degrade that was observed in [2] and achieve similar results. Thus, we could just remove `ecf_optim` and `snd_cwnd_before_idle_restart`.

> I'm hesitant to introduce this kind of change to the TCP-stack without being convinced that it is really necessary.

I can absolutely understand your view and I also do not want to modify the TCP-stack at all. I was strongly encouraged by @teto and @matbee to share the ECF version that I (somehow) ported to a current version so far. Unfortunately, I got my hands on the original ECF implementation so late, that it didn't made it into my evaluation.

I will update this pull-request with a more streamlined version of ECF matching the other schedulers. Until now I sticked very close to the orginal code published for v0.89

[1] https://dl.acm.org/citation.cfm?doid=3078505.3078552
[2] https://dl.acm.org/citation.cfm?doid=3143361.3143376